docs/RFCS/typing.md

```diff
@@ -552,6 +552,33 @@ being type checked, it is up to the caller to assert a specific type
 is returned from the expression typing and throw a type checking error
 if necessary.
 
+A desired type will fit into one of three classes: fully-specified types,
+partially-specified types, or unspecified (wildcard) types. Fully-specified types
+are types where the base type and all parameterized types (if applicable) are SQL
+types. Partially-specified types are types where the base type is a SQL type, but 
+one or more parameterized types is a wildcard. Unspecified types are where the base 
+class itself is a wildcard. The table below gives examples of each.
+
+| desired type class     | examples
+|:----------------------:|:--------------------------
+| fully-specified        | string, int, array\<decimal\>, tuple\<int, string\>
+| partially-specified    | array\<\*\>, tuple\<\*, \*\>, array\<decimal\>, tuple\<\*, string\>
+| unspecified (wildcard) | \*
+
+We say that two type are "equivalent" when either of the following statements
+are true:
+
+- they having identical base types and have equivalent parameterized types (if applicable)
+- one or both of the types is unspecified
+
```

Comment:
You can simplify this as follows: We say that two desired type patterns are equivalent when they are structurally equivalent given that a wildcard is equivalent to any type.